lib/Doctrine/ORM/AbstractQuery.php

```diff
@@ -1131,14 +1135,21 @@ public function iterate($parameters = null, $hydrationMode = null)
      */
     public function toIterable(iterable $parameters = [], $hydrationMode = null): iterable
     {
+        if ($parameters instanceof ArrayCollection) {
+            Deprecation::trigger(
+                'doctrine/orm',
+                'https://github.com/doctrine/orm/pull/9816',
+                'Passing an ArrayCollection to %s is deprecated.',
+                __METHOD__
+            );
+            $parameters = $parameters->toArray();
+        }
+
         if ($hydrationMode !== null) {
             $this->setHydrationMode($hydrationMode);
         }
 
-        if (
-            ($this->isCountable($parameters) && count($parameters) !== 0)
-            || ($parameters instanceof Traversable && iterator_count($parameters) !== 0)
-        ) {
+        if (count($parameters) !== 0) {
```

Comment:
What do you think about that first commit? Too dangerous?

Reply:
It can definitely break something after an update to 2.13 because it was able to inject other iterators into `toIterable()`. Maybe this PR is more than just deprecating `ArrayCollection`.

Reply:
OK, I went the safe road: I widened the phpdoc to `iterable`. This complex check can still go because I'm converting `Traversable` to array with `iterator_to_array`, so as not to trigger deprecations when calling other methods.

Reply:
`iterator_to_array` will remove all keys in its current usage. As far as I understand they need to be kept in the resulting array..

Reply:
`iterator_to_array()` should actually preserve keys. https://3v4l.org/Xo4AI

Reply:
I mixed up the default value of the second parameter.